File: src/main/java/loxal/epvin/datastore/model/OfyLocator.java
/*
 * Copyright 2012 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package loxal.epvin.datastore.model;

import com.google.web.bindery.requestfactory.shared.Locator;
import com.googlecode.objectify.Objectify;
import com.googlecode.objectify.ObjectifyService;

public class OfyLocator extends Locator<DatastoreEntity, Long> {
    @Override
    public DatastoreEntity create(final Class<? extends DatastoreEntity> cls) {
        try {
            return cls.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public DatastoreEntity find(final Class<? extends DatastoreEntity> cls, final Long id) {
        final Objectify ofy = ObjectifyService.ofy();
        return ofy.load().type(cls).id(id).now();
    }

    // never called
    @Override
    public Class<DatastoreEntity> getDomainType() {
        return DatastoreEntity.class; // could also return null
    }

    @Override
    public Long getId(final DatastoreEntity domainEntity) {
        return domainEntity.getId();
    }

    @Override
    public Class<Long> getIdType() {
        return Long.class;
    }

    @Override
    public Object getVersion(final DatastoreEntity domainEntity) {
        return domainEntity.getVersion();
    }
}

File: src/main/kotlin/loxal/epvin/datastore/model/OfyLocator.kt
/*
 * Copyright 2012 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package loxal.epvin.datastore.model

import com.google.web.bindery.requestfactory.shared.Locator
import com.googlecode.objectify.Objectify
import com.googlecode.objectify.ObjectifyService

public class OfyLocator : Locator<DatastoreEntity, Long>() {
    override fun create(cls: Class<out DatastoreEntity>): DatastoreEntity {
        try {
            return cls.newInstance()
        } catch (e: InstantiationException) {
            throw RuntimeException(e)
        } catch (e: IllegalAccessException) {
            throw RuntimeException(e)
        }

    }

    override fun find(cls: Class<out DatastoreEntity>, id: Long): DatastoreEntity {
        val ofy = ObjectifyService.ofy()
//        return ofy.load().type(cls).id(id).now()  // TODO fix this and do not use the expression below
        return DatastoreEntity()
    }

    // never called
    override fun getDomainType(): Class<DatastoreEntity> {
        return javaClass<DatastoreEntity>() // could also return null
    }

    override fun getId(domainEntity: DatastoreEntity): Long? {
        return domainEntity.id
    }

    override fun getIdType(): Class<Long> {
        return javaClass<Long>()
    }

    override fun getVersion(domainEntity: DatastoreEntity): Any {
        return domainEntity.version
    }
}
